minetest/minetest
# Proposal: API to control shadow intensity from the game/mod

This is a proposal to fix #11365 and #11972.

## Specification

*Lua API*

A game should be able to control intensity of ambient shadows as part of a new set_lighting() API call
```
   local player = core.get_player_by_name("singleplayer")
   player:set_lighting( { shadows = { intensity = 0.5 } } )
```

set_lighting() API call will also be used for ambient brightness (#11499) and saturation (#11854). A matching get_lighting() API call will be used to retrieve the current values of the lighting and shadows parameters.

*Network*
set_lighting call translates into a new TOCLIENT_SET_LIGHTING network packet containing parameters packed one after another.

*Client*

Introduce new setting `shadow_intensity_gamma` to control the visual darkness of shadow intensity values to accommodate for differences between computer screens.

The formula is: actual_intensity = (input_intensity) * (1 / shadow_intensity_gamma).

When actual_intensity is zero:
* Stop generating shadow mapping textures
* Disable shadow mapping code in the shaders.

## To do

This PR is Ready for Review.

Follow-up tasks:
- [x] Slight change to the Lua API, [this](https://irc.minetest.net/minetest-dev/2022-01-22#i_5926013)

## How to test

* Enable dynamic shadows
* Tune `shadow_strength_gamma` parameter
* Launch a devtest world
* `/set_lighting 0` - shadows disappear, and FPS goes up
* `/set_lighting 0.5` - shadows appear at half intensity
* `/set_lighting 1` - shadows appear at full intensity (black)

Check `games/devtest/mods/experimental/lighting.lua` for the example code.

## Note

Deleting and recreating the SM textures at runtime causes irrlicht errors and an eventual crash, won't fix in this PR.

Edit: spelling, add a todo
